### ISSUE TITLE
DecoderNLS.Convert doesn't always out the correct value for 'completed'

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/coreclr#27191]
This is pretty much the same issue as https://github.com/dotnet/coreclr/issues/23020, but we forgot to apply the same fix to the `DecoderNLS` class.

Repro code:

```cs
Encoder encoder = Encoding.UTF8.GetEncoder();
encoder.Convert(new char[] { '\ud800' }, 0, 1, new byte[0], 0, 0, flush: false, out int charsUsed, out int bytesUsed, out bool completed);
Console.WriteLine(charsUsed);
Console.WriteLine(bytesUsed);
Console.WriteLine(completed);
Console.WriteLine();

Decoder decoder = Encoding.UTF8.GetDecoder();
decoder.Convert(new byte[] { 0xF0 }, 0, 1, new char[0], 0, 0, flush: false, out bytesUsed, out charsUsed, out completed);
Console.WriteLine(bytesUsed);
Console.WriteLine(charsUsed);
Console.WriteLine(completed);
```

Output on .NET Framework 4.8:

```txt
1
0
True     <-- this is incorrect

1
0
True     <-- this is incorrect
```

Output on .NET Core 3.0:

```txt
1
0
False    <-- this is correct

1
0
True     <-- this is still incorrect
```

/cc @tarekgh as FYI

**Comment by tarekgh** [comment]
@GrabYourPitchforks do you think this could be important enough to port to 3.1?

**Comment by GrabYourPitchforks** [comment]
Unsure. It's affecting my ability to create unit tests for https://github.com/dotnet/corefx/issues/41326. But I seem to be the only person in the world running into these issues. :)

**Comment by tarekgh** [comment]
@GrabYourPitchforks this make it very important to port it then :-)

**Comment by GrabYourPitchforks** [comment]
Eh, sure. The worst shiproom can say is "no". :)

**Comment by ahsonkhan** [comment]
Re-opening until it's fixed in 3.1 (or we change the milestone).
